// @formio/premium/lib/components/Tagpad/editForm/Tagpad.edit.display.js
export default [
    {
        type: 'textfield',
        label: 'Image Url',
        input: true,
        key: 'imageUrl',
        weight: 20,
    },
    {
        type: 'select',
        input: true,
        key: 'imageType',
        label: 'Image Type',
        tooltip: 'Element used to display background image',
        weight: 20,
        defaultValue: 'image',
        dataSrc: 'values',
        data: {
            values: [
                { label: 'Image', value: 'image' },
                { label: 'SVG', value: 'svg' },
            ],
        },
    },
    {
        type: 'textfield',
        label: 'Width',
        input: true,
        key: 'canvasWidth',
        weight: 20,
    },
    {
        type: 'textfield',
        label: 'Height',
        input: true,
        key: 'canvasHeight',
        weight: 20,
    },
    {
        type: 'textfield',
        label: 'Dot Size',
        input: true,
        key: 'dotSize',
        placeholder: '10',
        weight: 20,
    },
    {
        type: 'textfield',
        label: 'Dot Stroke Size',
        input: true,
        key: 'dotStrokeSize',
        placeholder: '2',
        weight: 20,
    },
    {
        type: 'textfield',
        label: 'Dot Stroke Color',
        input: true,
        key: 'dotStrokeColor',
        placeholder: '#333',
        weight: 20,
    },
    {
        type: 'textfield',
        label: 'Dot Fill Color',
        input: true,
        key: 'dotFillColor',
        placeholder: '#ccc',
        weight: 20,
    },
];
